### PR TITLE
Alignment of obs/dyn time steps and refactored forward simulation

### PR DESCRIPTION
**Pull request by THargreaves**
Changes:

- Modified the definitions of the interface to align the obs/dyn time steps as per #54 
- Introduced type parameters to the dyn/obs representing the state/observation types
- Used these type parameters and new alignment to refactor the forward simulation code
- Added a unit test for forward simulation

I have a fear that these new type parameters, although useful, might be a bit over-engineered and lead to sticky edge-cases. Very happy to revert those changes if others think that is best.

TODO:

- [ ] If we're happy with these type parameters, it may be worth updating the method signatures to include them, e.g. `distribution(dyn::LatentDynamics, step::Integer, prev_state, extra)` -> `distribution(dyn::LatentDynamics{T}, step::Integer, prev_state:T, extra)`. This could potentially lead to method ambiguity errors if the user misses out these types in their own definition (their dynamics will be more specific, but the documentation's state will be more specific).
- [ ] I'm currently not returning the initial state `x0` or allowing an `extra` to be passed to initialisation. This is because doing so would required `xs`/`extras` to have a zero index. This could be handled by offset arrays but that feels like a big dependency. `Kalman.jl` handles this by using a vector of pairs `t => x` but that feels clunky. I question how much the initial state/`extra` for initial are actually needed.

### CODE REVIEW

File: src/SSMProblems.jl [hunk 1]
```diff
@@ -31,11 +42,25 @@ abstract type LatentDynamics end
...
 """
-abstract type ObservationProcess end
+    eltype(obs::ObservationProcess)
+
+    Return a pair of types for the state and observation of the observation process.
+
+    The first type is the type of the state of the latent dynamics, and the second type is
+    the type of the observation.
+"""
+Base.eltype(::Type{<:ObservationProcess{T,U}}) where {T,U} = (T, U)
```
Comment (charlesknipp) [inline src/SSMProblems.jl:60]: I'm not too crazy about having both the type of the state and observation, especially since the observation type should already known when calling a filtering algorithm given that the user's model is type stable.

I would be curious to see of any cases in which this type parameter gets used. Maybe I'm not thinking creatively enough, and it would be good to have.

File: src/SSMProblems.jl [hunk 2]
```diff
@@ -31,11 +42,25 @@ abstract type LatentDynamics end
...
 
 """
-    distribution(dyn::LatentDynamics, extra)
```
Comment (charlesknipp) [inline src/SSMProblems.jl:38]: For cases like trend dynamics for an $I(1)$ series where $y_{1}$ is sufficiently non-zero, I feel like `extra` might be of use here. We could always define the initial value in the `LatentDynamic` but that feels redundant since we observe $y_{1}$ in the filtering code already. Tossing it into the `extra` argument eliminates that problem, but maybe I'm in the minority here.

**Review by charlesknipp — COMMENTED** [review]
Besides some minor comments, I like what I see.

File: src/SSMProblems.jl (hunk 1 above)
Comment (THargreaves) [reply src/SSMProblems.jl:60]: That's an interesting point. 

The current use-case for the observation type is creating the container to put samples in during forward simulation. I.e. `ys = Vector{eltype(model.obs)}(undef, T)`.

As for the state type, this is currently only used for validation. When creating an SSM, a check is made to ensure that the dynamics and observations are expecting the same type for the state. I'm not sure if that's a good idea though—is there anything wrong with the LatentDynamics being Float32 and then casting these to Float64 for the observations, for example.

File: src/SSMProblems.jl (hunk 2 above)
Comment (THargreaves) [reply src/SSMProblems.jl:38]: I'm not familiar with the $I(1)$ series. Are you saying this is an example where it's useful to have a "control" variable for the initialisation of X_0?

File: src/SSMProblems.jl (hunk 1 above)
Comment (FredericWantiez) [reply src/SSMProblems.jl:60]: I would think that you would get the observation type from the data at this point, most likely your model specializes based on the data you get. Don't have a strong opinion against having both though

Comment (THargreaves) [reply src/SSMProblems.jl:60]: Very much agree that that is the case when filtering, but what about when forward simulating from the model?

I guess you can always do the whole

```
y1 = simulate(model.obs, ...)
ys = Vector{typeof(y1}}(undef, T)
ys[1] = 1
```

Bit clunky, but so is forcing the user to define the observation type upfront...

Comment (charlesknipp) [reply src/SSMProblems.jl:60]: > I would think that you would get the observation type from the data at this point, most likely your model specializes based on the data you get.

@FredericWantiez this is exactly what I had in mind

> Very much agree that that is the case when filtering, but what about when forward simulating from the model?

@THargreaves this is definitely valid. Honestly, I'm not hellbent on this either way. I think we keep it for now, and if we lack any uses beyond DGPs, then we can always can it later.

File: src/SSMProblems.jl (hunk 2 above)
Comment (charlesknipp) [reply src/SSMProblems.jl:38]: @THargreaves its a time series notation for the [order of integration](https://en.wikipedia.org/wiki/Order_of_integration). It's a pretty common terminology in macroeconomics.

Anyways, I don't know if "control" is the right word to use here, but I think it may be a nice way to initialize an algorithm. The case I had in mind was running a filter for feature extraction (or signal decomposition) for a series that starts well above 0. If the model defaults to initialize around 0 and the first data point lies around 1000, then estimation looks messy.

**Comment by THargreaves** [comment]
All suggested changes made. Since this is a breaking change to the interface, I've bumped the minor version. Will merge now.
